File: Code/TDFDAWFS2DSG.m
function TDFDAWFS2DSG

% TDFDAWFS2DSG
% This is a program of Time Domain Finite Difference Acoustic Wave Field Simulating with 2-Dimension Staggered Grid.
% Written by Tche.L. from USTC, 2016,6.

clc; clear; close all;
% format long;

%% Input parameters

nx = 101;               % the number of grid nodes in x-direction.
nz = 101;               % the number of grid nodes in z-direction.
npmlz = 20;            	% the number of grid nodes in top and bottom side of PML absorbing boundary.
npmlx = 20;             % the number of grid nodes in left and right side of PML absorbing boudary.
sx = 50;                % the grid node number of source position in x-direction.
sz = 50;                % the grid node number of source position in z-direction.
dx = 5;                 % the grid node interval in x-direction; Unit: m.
dz = 5;                 % the grid node interval in z-direction; Unit: m.
nt = 500;               % the number of time nodes for wave calculating.
dt = 1e-3;            	% the time node interval; Unit: s.
nppw = 12;              % the node point number per wavelength for dominant frequency of Ricker wavelet source.
ampl = 1.0e0;           % the amplitude of source wavelet.
xrcvr = 1:3:nx;         % the grid node number in x-direction of reciver position on ground.
nodr = 3;               % half of the order number for spatial difference.

%% Determine the difference coefficients

B = [1 zeros(1,nodr - 1)]';
A = NaN*ones(nodr,nodr);
for i = 1:1:nodr
    A(i,:) = (1:2:2*nodr - 1).^(2*i - 1);
end
C = A\B;

%% Model and source

Nz = nz + 2*npmlz;
Nx = nx + 2*npmlx;

vp = 2000*ones(Nz,Nx);                                                      % the velocity of acoustic wave of model; Unit: m/s.
rho = 1000*ones(Nz,Nx);                                                     % the density of model; Unit: kg/m^3.
rho(fix(Nz/3):end,fix(Nx/2):end) = 500;
vp(fix(Nz/3):end,fix(Nx/2):end) = 1000;

f0 = min(vp(:))/(min(dx,dz)*nppw);                                          % the dominant frequency of source Ricker wavelet; Unit: Hz.
t0 = 1/f0;                                                                  % the time shift of source Ricker wavelet; Unit: s; Suggest: 0.02 if fm = 50, or 0.05 if fm = 20.
t = dt*(1:1:nt);
src = (1 - 2*(pi*f0.*(t - t0)).^2).*exp( - (pi*f0*(t - t0)).^2);           	% the time series of source wavelet.
% The source wavelet formula refers to the equations (18) of Collino and Tsogka, 2001.

%% Perfectly matched layer absorbing factor

% R = 1e-6;                                                                   % Recommend: $R = 1e-2, if $nabsr = 5; $R = 1e-3, if $nabsr = 10; $R = 1e-4, if $absr = 20.
% dpml0z = log(1/R)*3*max(vp(:))/(2*npmlz);
dpml0z = 3*max(vp(:))/dz*(8/15 - 3/100*npmlz + 1/1500*npmlz^2);
dpmlz = zeros(Nz,Nx);
dpmlz(1:npmlz,:) = (dpml0z*((npmlz: - 1:1)./npmlz).^2)'*ones(1,Nx);
dpmlz(npmlz + nz + 1:Nz,:) = dpmlz(npmlz: - 1:1,:);
dpml0x = 3*max(vp(:))/dx*(8/15 - 3/100*npmlx + 1/1500*npmlx^2);
dpmlx = zeros(Nz,Nx);
dpmlx(:,1:npmlx) = ones(Nz,1)*(dpml0x*((npmlx: - 1:1)./npmlx).^2);
dpmlx(:,npmlx + nx + 1:Nx) = dpmlx(:,npmlx: - 1:1);
% The PLM formula refers to the equations (2) and (3) of Marcinkovich and Olsen, 2003.

%% Wavefield calculating

rho1 = rho;             % or = [(rho(:,1:end - 1) + rho(:,2:end))./2 (2*rho(:,end) - rho(:,end - 1))];
rho2 = rho;             % or = [(rho(1:end - 1,:) + rho(2:end,:))./2; (2*rho(end,:) - rho(end - 1,:))];

Coeffi1 = (2 - dt.*dpmlx)./(2 + dt.*dpmlx);
Coeffi2 = (2 - dt.*dpmlz)./(2 + dt.*dpmlz);
Coeffi3 = 1./rho1./dx.*(2*dt./(2 + dt.*dpmlx));
Coeffi4 = 1./rho2./dz.*(2*dt./(2 + dt.*dpmlz));
Coeffi5 = rho.*(vp.^2)./dx.*(2*dt./(2 + dt.*dpmlx));
Coeffi6 = rho.*(vp.^2)./dz.*(2*dt./(2 + dt.*dpmlz));

% +++++++++++++++++++++++++++++++++++++ approximate coeffient ++++++++++++++++++++++++++++++++++++++
% Coeffi1 = 1 - dt.*dpmlx;
% Coeffi2 = 1 - dt.*dpmlz;
% Coeffi3 = 1./rho./dx.*dt;
% Coeffi4 = 1./rho./dz.*dt;
% Coeffi5 = rho.*(vp.^2)./dx.*dt;
% Coeffi6 = rho.*(vp.^2)./dz.*dt;
% --------------------------------------------------------------------------------------------------

NZ = Nz + 2*nodr;                                                           % All values of the outermost some columns are set to zero to be a boundary condition: all of wavefield values beyond the left and right boundary are null.
NX = Nx + 2*nodr;                                                           % All values of the outermost some rows are set to zero to be a boundary condition: all of wavefield values beyond the top and bottom boundary are null.

Znodes = nodr + 1:NZ - nodr;
Xnodes = nodr + 1:NX - nodr;
znodes = nodr + npmlz + 1:nodr + npmlz + nz;
xnodes = nodr + npmlx + 1:nodr + npmlx + nx;
nsrcz = nodr + npmlz + sz;
nsrcx = nodr + npmlx + sx;

Ut = NaN*ones(NZ,NX,2);                                                     % the wavefield value preallocation.
Uz = zeros(NZ,NX,2);                                                        % The initial condition: all of wavefield values are null before source excitation.
Ux = zeros(NZ,NX,2);                                                        % The initial condition: all of wavefield values are null before source excitation.
Vz = zeros(NZ,NX,2);                                                        % The initial condition: all of wavefield values are null before source excitation.
Vx = zeros(NZ,NX,2);                                                        % The initial condition: all of wavefield values are null before source excitation.
Psum = NaN*ones(Nz,Nx);

U = NaN*ones(nz,nx,nt);

tic;
for it = 1:1:nt
    fprintf('The calculating time node is: it = %d\n',it);
    Ux(nsrcz,nsrcx,1) = Ux(nsrcz,nsrcx,1) + ampl*src(it)./2;
    Uz(nsrcz,nsrcx,1) = Uz(nsrcz,nsrcx,1) + ampl*src(it)./2;
    Ut(:,:,1) = Ux(:,:,1) + Uz(:,:,1);
    U(:,:,it) = Ut(znodes,xnodes,1);
    Psum(:,:) = 0;
    for i = 1:1:nodr
        Psum = Psum + C(i).*(Ut(Znodes,Xnodes + i,1) - Ut(Znodes,Xnodes + 1 - i,1));
    end
    Vx(Znodes,Xnodes,2) = Coeffi1.*Vx(Znodes,Xnodes,1) - Coeffi3.*Psum;
    Psum(:,:) = 0;
    for i = 1:1:nodr
        Psum = Psum + C(i).*(Ut(Znodes + i,Xnodes,1) - Ut(Znodes + 1 - i,Xnodes,1));
    end
    Vz(Znodes,Xnodes,2) = Coeffi2.*Vz(Znodes,Xnodes,1) - Coeffi4.*Psum;
    Psum(:,:) = 0;
    for i = 1:1:nodr
        Psum = Psum + C(i).*(Vx(Znodes,Xnodes - 1 + i,2) - Vx(Znodes,Xnodes - i,2));
    end
    Ux(Znodes,Xnodes,2) = Coeffi1.*Ux(Znodes,Xnodes,1) - Coeffi5.*Psum;
    Psum(:,:) = 0;
    for i = 1:1:nodr
        Psum = Psum + C(i).*(Vz(Znodes - 1 + i,Xnodes,2) - Vz(Znodes - i,Xnodes,2));
    end
    Uz(Znodes,Xnodes,2) = Coeffi2.*Uz(Znodes,Xnodes,1) - Coeffi6.*Psum;
    Ut(:,:,1) = Ut(:,:,2);
    Uz(:,:,1) = Uz(:,:,2);
    Ux(:,:,1) = Ux(:,:,2);
    Vz(:,:,1) = Vz(:,:,2);
    Vx(:,:,1) = Vx(:,:,2);
end
toc;

%% Plotting

% Wavefield Snapshot
figure;% colormap gray;
clims = [min(U(:)) max(U(:))]./5;
for it = 1:5:nt
    imagesc((0:nx - 1).*dx,(0:nz - 1).*dz,U(:,:,it));%,clims);
    set(gca,'xaxislocation','top'); axis equal; axis([0 (nx - 1)*dx 0 (nz - 1)*dz]);
    colorbar; xlabel('x distance (m)'); ylabel('z depth (m)');
    title(sprintf('the snapshot of %.1f ms',it*dt*1e3),'position',[(nx - 1)*dx/2,(nz - 1)*dz*(1 + 0.07)]);
    pause(0.01);
end

% Synthetic Seismogram
syngram(:,:) = U(1,xrcvr,:);
synmax = max(abs(syngram(:)));
rcvrintv = (xrcvr(2) - xrcvr(1))*dx;
syngram = syngram./synmax.*(rcvrintv/2);
[nsyn,~] = size(syngram);
figure; hold on;
for i = 1:1:nsyn
    plot(syngram(i,:) + (xrcvr(i) - 1)*dx,t.*1e3);
end
xlabel('x distance (m)'); ylabel('travel time (ms)');
title('Synthetic Seismogram','position',[(xrcvr(1) + xrcvr(nsyn))*dx/2,t(end)*1e3*(1 + 0.07)]);
set(gca,'xaxislocation','top');
set(gca,'YDir','reverse');
hold off;

end

%% References

% Collino and Tsogka, 2001. Geophysics, Application of the perfectly matched absorbing layer model to the linear elastodynamic problem in anisotropic heterogeneous media.
% Marcinkovich and Olsen, 2003. Journal of Geophysical Research, On the implementation of perfectly mathced layers in a three-dimensional fourth-order velocity-stress finite difference scheme.

File: Code/TDFDAWFS2DSG.f90
MODULE InputPara

  IMPLICIT NONE

  PUBLIC
  INTEGER, PARAMETER :: nx = 101, nz = 101                      ! nx: the total number of grid nodes in x-direction; nz: the total number of grid nodes in z-direction.
  INTEGER, PARAMETER :: npmlx = 20, npmlz = 20                  ! npmlx: the total number of grid nodes in top and bottom side of PML absorbing boundary; npmlz: the total number of grid nodes in left and right side of PML absorbing boundary.
  INTEGER, PARAMETER :: sx = 50, sz = 50                        ! sx: the grid node number of source position in x-direction; sz: the grid node number of source position in z-direction.
  INTEGER, PARAMETER :: dx = 5, dz = 5                          ! dx: the grid node interval in x-direction; dz: the grid node interval in z-direction; Unit: m.
  INTEGER, PARAMETER :: nt = 500                                ! the total number of time nodes for wave calculating.
  REAL   , PARAMETER :: dt = 1.0E-3                             ! the time node interval, Unit: s.
  INTEGER, PARAMETER :: nppw = 12                               ! the total node point number per wavelength for dominant frequency of Ricker wavelet source.
  REAL   , PARAMETER :: amp = 1.0E0                             ! the amplitude of source wavelet.
  INTEGER, PARAMETER :: nodr = 3                                ! half of the order number for spatial difference.
  INTEGER, PARAMETER :: irstr = 1                               ! the node ID of starting reciver point.
  INTEGER, PARAMETER :: nrintv = 3                              ! the total node number between each two adjacent recivers.
  INTEGER, PARAMETER :: itstr = 1                               ! the time node ID of the first snapshot.
  INTEGER, PARAMETER :: ntintv = 5                              ! the total time node number between each two followed snapshot.

  REAL    :: src(nt)                                            ! the time series of source wavelet.
  REAL    :: vp(nz, nx), rho(nz, nx)                            ! vp: the velocity of acoustic wave of model, Unit: m/s; rho: the density of model, Unit: kg/m^3.   
  INTEGER, PARAMETER :: nrcvr = CEILING(REAL(nx)/nrintv)        ! the total number of all recivers.
  INTEGER :: xrcvr(nrcvr)                                       ! the grid node number in x-direction of reciver position on ground.

  INTEGER, PRIVATE :: i
  
  PRIVATE nppw, amp
  PRIVATE ModelVpRho, SrcWavelet

  CONTAINS
    SUBROUTINE IntlzInputPara()
      xrcvr = [ (irstr + (i - 1)*nrintv, i = 1, nrcvr) ]
      CALL ModelVpRho()
      CALL SrcWavelet()
    END SUBROUTINE IntlzInputPara
    SUBROUTINE ModelVpRho()
      ! here you can reset $vp$ and $rho$ for the model.
      vp = 2000
      vp(nz/3:nz, nx/2:nx) = 1000
      rho = 1000
      rho(nz/3:nz, nx/2:nx) = 500
    END SUBROUTINE ModelVpRho
    SUBROUTINE SrcWavelet()
      ! here you can reset $src$ for the source wavelet.
      REAL :: f0, t0, pi = 3.1415926
      REAL :: t(nt)
      f0 = MINVAL(vp)/(MIN(dx, dz)*nppw)
      t0 = 1/f0
      t = [ (i*dt, i = 1, nt) ]
      src = amp*(1 - 2*(pi*f0*(t - t0))**2)*EXP( - (pi*f0*(t - t0))**2)
    END SUBROUTINE SrcWavelet

END MODULE InputPara    

MODULE WaveExtrp

  USE InputPara
  IMPLICIT NONE

  PRIVATE
  REAL :: C(nodr)                                               ! the difference coefficients of spatial the $2*nodr$-th order difference approximating.
  INTEGER, PARAMETER :: Nzz = nz + 2*npmlz, Nxx = nx + 2*npmlx  ! Nzz: the total number of grid nodes in z-direction of compute-updating zone including PML layer; Nxx: the total number of grid nodes in x-direction of compute-updating zone including PML layer.
  REAL :: vpp(Nzz, Nxx), rhoo(Nzz, Nxx)                         ! vpp: the velocity of the expanded model including PML layer, Unit: m/s; rhoo: the density of the expanded model including PML layer, Unit: kg/m^3.
  REAL :: dpmlz(Nzz, Nxx), dpmlx(Nzz, Nxx)                      ! dpmlz: the PML damping factor in z-direction; dpmlx: the PML damping factor in x-direction.
  REAL :: Coef1(Nzz, Nxx), Coef2(Nzz, Nxx), &
    & Coef3(Nzz, Nxx), Coef4(Nzz, Nxx), &
    & Coef5(Nzz, Nxx), Coef6(Nzz, Nxx)                          ! Coef1 ~ Coef6: the coefficients of wavefield time-extrapolating formula.

  INTEGER :: i, j

  REAL, PUBLIC :: P(nz, nx, nt)                                 ! the calculating wavefield component varying with time.

  PUBLIC WaveExec

  CONTAINS
    SUBROUTINE WaveExec()
      CALL CalC()
      CALL ModelExpand()
      CALL CalCoefs()
      CALL CalWave()
    END SUBROUTINE WaveExec
    SUBROUTINE CalC()
      REAL :: rtemp1, rtemp2
      DO i = 1,nodr,1
        rtemp1 = 1.0
        rtemp2 = 1.0
        DO j = 1,nodr,1
          IF(j == i) CYCLE
          rtemp1 = rtemp1*((2*j - 1)**2)
          rtemp2 = rtemp2*ABS((2*i - 1)**2 - (2*j - 1)**2)
        END DO
        C(i) = ( - 1)**(i + 1)*rtemp1/((2*i - 1)*rtemp2)
      END DO
    END SUBROUTINE CalC
    SUBROUTINE ModelExpand()
      vpp = 0.0
      rhoo = 0.0
      vpp(npmlz + 1:npmlz + nz, npmlx + 1:npmlx + nx) = vp
      rhoo(npmlz + 1:npmlz + nz, npmlx + 1:npmlx + nx) = rho
      DO i = 1,npmlx,1
        vpp(:, i) = vpp(:, npmlx + 1)
        vpp(:, npmlx + nx + i) = vpp(:, npmlx + nx)
        rhoo(:, i) = rhoo(:, npmlx + 1)
        rhoo(:, npmlx + nx + i) = rhoo(:, npmlx + nx)
      END DO
      DO i = 1,npmlz,1
        vpp(i, :) = vpp(npmlz + 1, :)
        vpp(npmlz + nz + i, :) = vpp(npmlz + nz, :)
        rhoo(i, :) = rhoo(npmlz + 1, :)
        rhoo(npmlz + nz + i, :) = rhoo(npmlz + nz, :)
      END DO
    END SUBROUTINE ModelExpand
    SUBROUTINE CalDpml()
      REAL :: dpml0z, dpml0x
      dpml0z = 3*MAXVAL(vp)/dz*(8.0/15 - 3.0/100*npmlz + 1.0/1500*(npmlz**2))
      DO i = 1,npmlz,1
        dpmlz(i, :) = dpml0z*((REAL(npmlz - i + 1)/npmlz)**2)
      END DO
      dpmlz(npmlz + nz + 1:Nzz, :) = dpmlz(npmlz:1:-1, :)
      dpml0x = 3*MAXVAL(vp)/dx*(8.0/15 - 3.0/100*npmlx + 1.0/1500*(npmlx**2))
      DO i = 1,npmlx,1
        dpmlx(:, i) = dpml0x*((REAL(npmlx - i + 1)/npmlx)**2)
      END DO
      dpmlx(:, npmlx + nx + 1:Nxx) = dpmlx(:, npmlx:1:-1)
    END SUBROUTINE CalDpml
    SUBROUTINE CalCoefs()
      CALL CalDpml()
      Coef1 = (2 - dt*dpmlx)/(2 + dt*dpmlx)
      Coef2 = (2 - dt*dpmlz)/(2 + dt*dpmlz)
      Coef3 = (2*dt/(2 + dt*dpmlx))/(rhoo*dx)
      Coef4 = (2*dt/(2 + dt*dpmlz))/(rhoo*dz)
      Coef5 = (2*dt/(2 + dt*dpmlx))*(rhoo*(vpp**2)/dx)
      Coef6 = (2*dt/(2 + dt*dpmlz))*(rhoo*(vpp**2)/dz)
    END SUBROUTINE CalCoefs
    SUBROUTINE CalWave()
      INTEGER :: it
      INTEGER, PARAMETER :: Nzzz = Nzz + 2*nodr, Nxxx = Nxx + 2*nodr
      INTEGER :: znds(nz) = [ (nodr + npmlz + i, i = 1,nz,1) ], &
        & xnds(nx) = [ (nodr + npmlx + i, i = 1,nx,1) ]
      INTEGER :: Zznds(Nzz) = [ (nodr + i, i = 1,Nzz,1) ], &
        & Xxnds(Nxx) = [ (nodr + i, i = 1,Nxx,1) ]
      INTEGER :: nsrcz = nodr + npmlz + sz, nsrcx = nodr + npmlx + sx
      REAL    :: Pt(Nzzz, Nxxx) = 0, &
        & Pz(Nzzz, Nxxx) = 0, Px(Nzzz, Nxxx) = 0, &
        & vz(Nzzz, Nxxx) = 0, vx(Nzzz, Nxxx) = 0
      REAL    :: SpcSum(Nzz, Nxx)
      DO it = 1,nt,1
        WRITE(*,"(A,G0)") 'The calculating time node is: it = ',it
        Px(nsrcz, nsrcx) = Px(nsrcz, nsrcx) + src(it)/2
        Pz(nsrcz, nsrcx) = Pz(nsrcz, nsrcx) + src(it)/2
        Pt(:, :) = Px(:, :) + Pz(:, :)
        P(:, :, it) = Pt(znds, xnds)
        SpcSum = 0
        DO i = 1,nodr,1
          SpcSum = SpcSum + C(i)*(Pt(Zznds, Xxnds + i) - Pt(Zznds, Xxnds + 1 - i))
        END DO
        vx(Zznds, Xxnds) = Coef1*vx(Zznds, Xxnds) - Coef3*SpcSum
        SpcSum = 0
        DO i = 1,nodr,1
          SpcSum = SpcSum + C(i)*(Pt(Zznds + i, Xxnds) - Pt(Zznds + 1 - i, Xxnds))
        END DO
        vz(Zznds, Xxnds) = Coef2*vz(Zznds, Xxnds) - Coef4*SpcSum
        SpcSum = 0
        DO i = 1,nodr,1
          SpcSum = SpcSum + C(i)*(vx(Zznds, Xxnds - 1 + i) - vx(Zznds, Xxnds - i))
        END DO
        Px(Zznds, Xxnds) = Coef1*Px(Zznds, Xxnds) - Coef5*SpcSum
        SpcSum = 0
        DO i = 1,nodr,1
          SpcSum = SpcSum + C(i)*(vz(Zznds - 1 + i, Xxnds) - vz(Zznds - i, Xxnds))
        END DO
        Pz(Zznds, Xxnds) = Coef2*Pz(Zznds, Xxnds) - Coef6*SpcSum
      END DO
    END SUBROUTINE CalWave

END MODULE WaveExtrp

!************************  TDFDAWFS2DSG  ***********************
! Time Domain Finite Difference Acoustic Wave Field Simulating with 2-Dimension Staggered Grid
! Written by Tche. L. from USTC, 2016.7.
! References:
!   Collino and Tsogka, 2001. Geophysics, Application of the perfectly matched absorbing layer model to the linear elastodynamic problem in anisotropic heterogeneous media.
!   Marcinkovich and Olsen, 2003. Journal of Geophysical Research, On the implementation of perfectly mathced layers in a three-dimensional fourth-order velocity-stress finite difference scheme.
!***************************************************************
PROGRAM TDFDAWFS2DSG

  USE InputPara
  USE WaveExtrp
  IMPLICIT NONE

  CHARACTER(LEN = 128) :: SnapFile = './data/Snapshot_****.dat'       ! the snapshot file name template.
  CHARACTER(LEN = 128) :: SyntFile = './data/SyntRcrd.dat'                       ! the synthetic record file name.
  REAL    :: SyntR(nrcvr, nt)
  INTEGER :: i

  CALL IntlzInputPara()
  CALL WaveExec()
  DO i = itstr,nt,ntintv
    WRITE(SnapFile(21:24),"(I4.4)") i
    CALL Output(TRIM(SnapFile), nz, nx, P(:, :, i))
  END DO
  DO i = 1,nt,1
    SyntR(:, i) = P(1, xrcvr, i)
  END DO
  CALL Output(TRIM(SyntFile), nrcvr, nt, SyntR)

END PROGRAM TDFDAWFS2DSG

SUBROUTINE Output(Outfile, M, N, OutA)
  IMPLICIT NONE
  CHARACTER(LEN = *), INTENT(IN) :: Outfile
  INTEGER, INTENT(IN) :: M, N
  REAL, INTENT(IN)    :: OutA(M, N)
  CHARACTER(LEN = 40) :: FmtStr
  INTEGER :: i, j
  INTEGER :: funit
  WRITE(FmtStr,"('(',G0,'E15.6)')") N
  OPEN(NEWUNIT = funit, FILE = Outfile, STATUS = 'UNKNOWN')
    DO i = 1,M,1
      WRITE(funit, FmtStr) (OutA(i, j), j = 1,N,1)
    END DO
  CLOSE(funit)
END SUBROUTINE Output
